asyncscope.md

```diff
@@ -783,29 +790,31 @@ In what context will the destructor run?
 Or we can teach that destructors will not block and indicate lifetime failures with `terminate()`.
 
-Software is less likely to ship with crashes and more likely to be fixed when there are crashes.
+One authors philosophy is that software is less likely to ship with crashes and more likely to be fixed when there are crashes.
 Deadlocks result in users forcefully terminating the app and forced terminations are rarely reported to the developer as a bug and even if reported, tend to have no debug data (stacks, dumps, etc...).
 If there is a lifetime bug that you want fixed -- it had better crash.
 
 Principles that lead to avoid blocking in the destructor:
 
-- Blocking must be explicit (exiting a sync scope is implicit -- and `shared_ptr` makes it even more scary).
+- Blocking must be explicit (exiting a sync scope is implicit -- and `shared_ptr` makes it even more scary as the destructor will potentially run at a different point each time).
 - Blocking must be grepable.
 - Blocking must be rare.
+- Blocking must be composable.
 - Blocking is like `reinterpret_cast<>` -- the name should be long and scary.
-- `join()` is a terrible design -- this is why async_scope has `on_empty()` instead.
-- Every asynchronous lifetime must be managed with non-blocking primitives and only `sync_wait()` is used to block.
+- `join()` is grepable and explicit, it is not rare, it is not composable (There is a separate blocking wait for each. One blocking wait for many different things to complete would be better)-- this is why async_scope has `when_empty()` instead.
+
+Every asynchronous operation must join with non-blocking primitives and only `sync_wait()` is used to block some composition of those primitives.
 
 ## Why doesn't the `async_scope` destructor stop all the nested and spawned senders?
 
 - `stop_callback` is not a destructor because:
   - `request_stop()` is **asking** for early completion.
   - `request_stop()` does not end the lifetime of the operation, `set_value()`, `set_error()` and `set_stopped()` end the lifetime -- those are the destructors for an operation.
-  - `request_stop()` might result in completion with `set_stopped()`, but `set_value()` and `set_error()` are equally likely.
+  - `request_stop()` might result in completion with `set_stopped()`, but `set_value()` and `set_error()` are equally valid.
 
 `request_stop()` should not be called from a destructor because:
 If a sync context intends to ask for early completion of an async operation, then it needs to wait for that operation to actually complete before continuing (`set_value()`, `set_error()` and `set_stopped()` are the destructors for the async operation), and sync destructors must not block.
-See [Why does `async_scope` terminate in the destructor instead of blocking like `jthread`?].
+_See_ [Why does `async_scope` terminate in the destructor instead of blocking like `jthread`?].
```

Comment:
That explanation is fine.. ..but it still leaves me wanting to have a ready-made async_scope wrapper that'll block until the scope is empty. That's likely to be somewhat common, and it eases the tension of (again) introducing something that's hard to use in an exception-safe manner without writing fair amounts of boilerplate. Sure, I expect quite many async_scopes to be class members, so it's not a huge burden to write the boilerplate in a destructor, but it's boilerplate that can be rather surprising for non-experts who can still survive with the default behavior of a wrapper that just sync_waits the scope. It sure took me a while to realize I need to do
    auto cleanupSender = async_scope.cleanup();
    unifex::sync_wait(cleanupSender);
in my orchestrator that spawns work using an async_scope.

Reply:
I am confused by this. The paper is explaining why every usage of an "async_scope wrapper that'll block until the scope is empty" is a bug. I may not understand your intent here

Reply:
I don't follow how they can all be bugs, when there seems to be valid use cases where your async 'controller' does exactly that, block until the scope is empty, by using sync_wait. All the wrapper does is move that code into its destructor. Whereas before, it's in the destructor of the controller.

Reply:
Furthermore, the paper explicitly says that the following is (of course) safe:
  async_scope s;
  s.spawn(snd);
  sync_wait(s.on_empty());
  That must mean that it's then also safe to write, instead
  async_scope s;
  async_scope_empty_guard sg(s);
  s.spawn(snd);
or
  async_scope_empty_guard s; // assuming that the guard will just create a scope inside
  s->spawn(snd);
because it does the same thing, the guard will just do sync_wait on s.on_empty(). The difference is that if something throws, the scope will still be waited upon.